File: react-native-totvs-clockin-vision/android/src/main/java/com/totvs/clockin/vision/utils/FileUtils.kt
package com.totvs.clockin.vision.utils

import android.content.Context
import android.os.Environment
import android.util.Base64
import android.util.Log
import androidx.annotation.RawRes
import androidx.annotation.RestrictTo
import com.totvs.clockin.vision.R
import com.totvs.clockin.vision.core.ClockInVisionModuleOptions
import java.io.BufferedInputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import kotlin.concurrent.thread

/**
 * Constants for the clock-in module
 */
internal object Constants {
    const val DEFAULT_MODEL_OUTPUT_DIR_NAME = "carol_offline_face_recognition"
    const val CAPTURES_OUTPUT_DIR_NAME = "images"
    const val PENDING_EMPLOYEES_OUTPUT_DIR_NAME = "pendingEmployeeImages"
    const val MODEL_SHAPE_DIR_NAME = "shape_predictor_5_face_landmarks.dat"
    const val MODEL_DESCRIPTOR_DIR_NAME = "dlib_face_recognition_resnet_model_v1.dat"
    const val NO_MEDIA = ".nomedia"
}

/**
 * Utility class to save transient state through utility changes
 */
private object TransientState {
    // the requested and desired model output directory name
    var modelOutputDirName: String? = null
}


/**
 * Get the by representation of the image in [path] if possible, null otherwise
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
internal fun getFileImageData(path: String): ByteArray? = try {
    val file = File(path)
    val bytes = ByteArray(file.length().toInt())
    BufferedInputStream(FileInputStream(file)).use { buffer ->
        buffer.read(bytes)
    }
    bytes
} catch (ex: Exception) {
    if (ClockInVisionModuleOptions.DEBUG_ENABLED) {
        Log.e("getFileImageData", "error reading image file at $path", ex)
    }
    null
}

/**
 * Get a base64 representation out of a [ByteArray] if possible, null otherwise
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
internal fun ByteArray.toBase64(): String? = if (size == 0) null else Base64.encodeToString(
    this,
    Base64.DEFAULT
)

/**
 * Get the base64 representation of an image if possible, null otherwise
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
internal fun getFileBase64(path: String): String? = getFileImageData(path)
    ?.let {
        it.toBase64()
    }


/**
 * Delete a file from the system, returns a boolean indicating the result of the operation.
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
internal fun deleteFile(path: String): Boolean = try {
    val file = File(path)
    if (file.exists()) {
        file.delete()
    } else {
        false
    }
} catch (ex: Exception) {
    if (ClockInVisionModuleOptions.DEBUG_ENABLED) {
        Log.e("deleteFile", "error deleting file at $path", ex)
    }
    false
}

/**
 * Copy a file from the raw folder to an specified file
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
internal fun copyRawFileOver(context: Context, @RawRes raw: Int, output: File) {
    try {
        context.resources.openRawResource(raw).use { input ->
            FileOutputStream(output).use { out ->
                val buffer = ByteArray(1024)
                var read: Int
                while (input.read(buffer).also { read = it } > 0) {
                    out.write(buffer, 0, read)
                }
            }
        }
    } catch (ex: Exception) {
        if (ClockInVisionModuleOptions.DEBUG_ENABLED) {
            Log.e("copyRawFileOver", "error copying raw file", ex)
        }
    }
}

/**
 * Set the model output directory name
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
internal fun setModelDirName(name: String) {
    TransientState.modelOutputDirName = name
}

/**
 * Returns the model output directory path
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
internal fun getModelOutputDir(): String {
    var dirName = if (TransientState.modelOutputDirName.isNullOrEmpty())
        Constants.DEFAULT_MODEL_OUTPUT_DIR_NAME
    else
        TransientState.modelOutputDirName

    val sdcard = Environment.getExternalStorageDirectory()
    return "${sdcard.absolutePath}${File.separator}$dirName"
}

/**
 * Returns the captures output directory path
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
internal fun getCapturesDirectory() =
    "${getModelOutputDir()}${File.separator}${Constants.CAPTURES_OUTPUT_DIR_NAME}"

/**
 * Returns the pending employees images output directory path
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
internal fun getPendingEmployeesDirectory() =
    "${getModelOutputDir()}${File.separator}${Constants.PENDING_EMPLOYEES_OUTPUT_DIR_NAME}"

/**
 * Returns the model shape output directory
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
internal fun getModelShapesDirectory() =
    "${getModelOutputDir()}${File.separator}${Constants.MODEL_SHAPE_DIR_NAME}"

/**
 * Returns the model descriptors output directory
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
internal fun getModelDescriptorsDirectory() =
    "${getModelOutputDir()}${File.separator}${Constants.MODEL_DESCRIPTOR_DIR_NAME}"


/**
 * Returns the .no_media directory path.
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
internal fun getNoMediaDirectory() =
    "${getModelOutputDir()}${File.separator}${Constants.NO_MEDIA}"

/**
 * Setup the model output directories
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
internal fun prepareModelDirectories(context: Context, onDone: () -> Unit) {
    thread {
        try {
            val model = File(getModelOutputDir())
            val media = File(getNoMediaDirectory())
            val images = File(getCapturesDirectory())
            val pending = File(getPendingEmployeesDirectory())
            val shapes = File(getModelShapesDirectory())
            val descriptors = File(getModelDescriptorsDirectory())

            model.mkdirs()
            media.mkdirs()
            images.mkdirs()
            pending.mkdirs()
            // If the file does not exist copy it over
            if (!shapes.exists()) {
                copyRawFileOver(context, R.raw.shape_predictor_5_face_landmarks, shapes)
            }
            if (!descriptors.exists()) {
                copyRawFileOver(context, R.raw.dlib_face_recognition_resnet_model_v1, descriptors)
            }
        } catch (ex: Exception) {
            if (ClockInVisionModuleOptions.DEBUG_ENABLED) {
                Log.e("prepareModelDirectories", "error configuring model directories")
            }
        }
        onDone()
    }
}

File: react-native/react-native-totvs-clockin-vision/ios/Utils/FileUtils.swift
//
//  FileUtils.swift
//  TOTVSClockInVision
//
//  Created by Jansel Rodriguez on 7/27/20.
//  Copyright © 2020 Facebook. All rights reserved.
//

import Foundation

fileprivate let DEFAULT_MODEL_OUTPUT_DIR_NAME = "carol_offline_face_recognition"
fileprivate let CAPTURES_OUTPUT_DIR_NAME = "images"
fileprivate let PENDING_EMPLOYEES_OUTPUT_DIR_NAME = "pendingEmployeeImages"
fileprivate let MODEL_SHAPE_NAME = "shape_predictor_5_face_landmarks"
fileprivate let MODEL_SHAPE_FILE_NAME = "shape_predictor_5_face_landmarks.dat"
fileprivate let MODEL_DESCRIPTOR_NAME = "dlib_face_recognition_resnet_model_v1"
fileprivate let MODEL_DESCRIPTOR_FILE_NAME = "dlib_face_recognition_resnet_model_v1.dat"
fileprivate let NO_MEDIA = ".nomedia"

/**
* Utility class to save transient state through utility changes
*/
fileprivate class TransientState {
    // the requested and desired model output directory name
    var modelOutputDirName: String? = nil
}

/**
* object representing the state if the lib constants
*/
fileprivate let transientState = TransientState()


func getFileBase64(atPath path: String) -> String? {
    if FileManager.default.fileExists(atPath: path) {
        let url = URL(fileURLWithPath: path)
        return url.dataRepresentation.base64EncodedString()
    }
    return nil
}

func deleteFile(atPath path: String) -> Bool {
    if FileManager.default.fileExists(atPath: path) {
        do {
            try FileManager.default.removeItem(atPath: path)
        } catch let error {
            print("deleteFile: error deleting file \(path): \(error)")
            return false
        }
        return true
    }
    return false
}

/**
* Set the model output directory name
*/
func setModelDirName(name: String) {
    transientState.modelOutputDirName = name
}

/**
* Returns the model output directory path
*/
func getModelOutputDir() -> String {
    return getModelOutputDirectoryURL().path
}

/**
* Returns the model output directory path convenientely as URL object.
*/
fileprivate func getModelOutputDirectoryURL() -> URL {
    var dirName: String
    if (nil == transientState.modelOutputDirName || transientState.modelOutputDirName!.isEmpty) {
        dirName = DEFAULT_MODEL_OUTPUT_DIR_NAME
    } else {
        dirName = transientState.modelOutputDirName!
    }
    return getDocumentsDirectory()
        .appendingPathComponent(dirName)
}

/**
* Returns the .no_media directory path.
*/
fileprivate func getNoMediaDirectoryURL() -> URL {
    return getModelOutputDirectoryURL()
        .appendingPathComponent(NO_MEDIA)
}

/**
* Returns the captures output directory path
*/
func getCapturesDirectoryURL() -> URL {
    return getModelOutputDirectoryURL()
        .appendingPathComponent(CAPTURES_OUTPUT_DIR_NAME)
}

/**
* Returns the pending employees images output directory path
*/
fileprivate func getPendingEmployeesDirectoryURL() -> URL {
    return getModelOutputDirectoryURL()
        .appendingPathComponent(PENDING_EMPLOYEES_OUTPUT_DIR_NAME)
}

/**
* Returns the model shape output directory
*/
fileprivate func getModelShapesFileURL() -> URL {
    return getModelOutputDirectoryURL()
        .appendingPathComponent(MODEL_SHAPE_FILE_NAME)
}

/**
* Returns the model descriptors output directory
*/
fileprivate func getModelDescriptorsURL() -> URL {
    return getModelOutputDirectoryURL()
        .appendingPathComponent(MODEL_DESCRIPTOR_FILE_NAME)
}

/**
* Setup the model output directories
*/
func prepareModelDirectories(onDone: @escaping (Bool) -> Void) {
    DispatchQueue.global().async {
        
        let model = getModelOutputDirectoryURL()
        let media = getNoMediaDirectoryURL()
        let images = getCapturesDirectoryURL()
        let pending = getPendingEmployeesDirectoryURL()
        let shapes = getModelShapesFileURL()
        let descriptor = getModelDescriptorsURL()

        let directories = [model, media, images, pending]
        do {
            try directories.forEach { d in
                try FileManager.default.createDirectory(at: d, withIntermediateDirectories: true, attributes: nil)
            }
            let (shapesResource, descriptorResource) = getModelResourceURLs()

            if !FileManager.default.fileExists(atPath: shapes.path) {
                try! FileManager.default.copyItem(at: shapesResource, to: shapes)
            }
            if !FileManager.default.fileExists(atPath: descriptor.path) {
                try! FileManager.default.copyItem(at: descriptorResource, to: descriptor)
            }

            DELETE_ME_AND_ALL_FILE_NAMED_HERE_AFTER_USING_ME___THIS_IS_FOR_TEST_ONLY()
        } catch let error {
            print("prepareModelDirectories: error configuring model directories: \(error)")
            return onDone(false)
        }
        onDone(true)
    }
}

fileprivate func DELETE_ME_AND_ALL_FILE_NAMED_HERE_AFTER_USING_ME___THIS_IS_FOR_TEST_ONLY() {
    // this function is used only for test purpose of copying the model descriptors file
    // to test the recognition model. it must be deleted after test.
    
    func getResourcesURLs() -> (deploy: URL, descJittered: URL, namesJiterred: URL, caffeModel: URL) {
        let frameworkBundle = Bundle(for: VisionModule.self)
        let moduleBundleURL = frameworkBundle.resourceURL?.appendingPathComponent("ClockInVision.bundle")
        
        guard let bundle = Bundle(url: moduleBundleURL!) else {
            fatalError(
                "ClockInVision.bundle not found, please check the file exists on react-native-totvs-clockin-vision package"
            )
        }
        let deploy = bundle.url(forResource: "deploy_prototxt", withExtension: "txt")!
        let descJittered = bundle.url(forResource: "face_descriptors_jittered", withExtension: "dat")!
        let namesJiterred = bundle.url(forResource: "face_names_jittered", withExtension: "dat")!
        let caffeModel = bundle.url(forResource: "res10_300x300_ssd_iter_140000", withExtension: "caffemodel")!
        return (deploy, descJittered, namesJiterred, caffeModel)
    }
    
    let (deploy, descJittered, namesJiterred, caffeModel) = getResourcesURLs()
    
    let deployDest = getModelOutputDirectoryURL().appendingPathComponent("deploy_prototxt.txt")
    let descJitteredDest = getModelOutputDirectoryURL().appendingPathComponent("face_descriptors_jittered.dat")
    let namesJiterredDest = getModelOutputDirectoryURL().appendingPathComponent("face_names_jittered.dat")
    let caffeModelDest = getModelOutputDirectoryURL().appendingPathComponent("res10_300x300_ssd_iter_140000.caffemodel")
        
    if !FileManager.default.fileExists(atPath: deployDest.path) {
        try! FileManager.default.copyItem(at: deploy, to: deployDest)
    }
    if !FileManager.default.fileExists(atPath: descJitteredDest.path) {
        try! FileManager.default.copyItem(at: descJittered, to: descJitteredDest)
    }
    if !FileManager.default.fileExists(atPath: namesJiterredDest.path) {
        try! FileManager.default.copyItem(at: namesJiterred, to: namesJiterredDest)
    }
    if !FileManager.default.fileExists(atPath: caffeModelDest.path) {
        try! FileManager.default.copyItem(at: caffeModel, to: caffeModelDest)
    }
}

/**
 Get all posible user app-specific document directories and returns the first one.
 */
fileprivate func getDocumentsDirectory() -> URL {
    let path = try! FileManager.default.url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
    return path
}

/**
 Returns the local model resources required for the face recognition model
 */
fileprivate func getModelResourceURLs() -> (shapesResource: URL, descriptorResource: URL) {
    let frameworkBundle = Bundle(for: VisionModule.self)
    let moduleBundleURL = frameworkBundle.resourceURL?.appendingPathComponent("ClockInVision.bundle")
    
    guard let bundle = Bundle(url: moduleBundleURL!) else {
        fatalError(
            "ClockInVision.bundle not found, please check the file exists on react-native-totvs-clockin-vision package"
        )
    }
    guard let shapes = bundle.url(forResource: MODEL_SHAPE_NAME, withExtension: "dat") else {
        fatalError(
            "\(MODEL_SHAPE_NAME) not found, please check the file exists on react-native-totvs-clockin-vision package"
        )
    }
    guard let descriptor = bundle.url(forResource: MODEL_DESCRIPTOR_NAME, withExtension: "dat") else {
        fatalError(
            "\(MODEL_SHAPE_NAME) not found, please check the file exists on react-native-totvs-clockin-vision package"
        )
    }
    return (shapes, descriptor)
}
